### PR TITLE
(draft) colpali WIP

### PR DESCRIPTION
**Pull request by I8dNLo**
Colpali draft
Works for texts, does not work for images

### CODE REVIEW

File: fastembed/text/colpali_model.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+class ColpaliTextModel(OnnxTextEmbedding):
+    query_prefix = "Query: "
+    bos_token = "<s>"
+    pad_token = "<pad>"
+
+    def _preprocess_onnx_input(
+        self, onnx_input: Dict[str, np.ndarray], **kwargs
+    ) -> Dict[str, np.ndarray]:
+        empty_image_placeholder = np.zeros((3, 448, 448), dtype=np.float32)
```
Comment (joein) [inline fastembed/text/colpali_model.py:37]: (3, 448, 448) is a magic combination required by particular models

File: fastembed/text/colpali_model.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    def onnx_embed(
+        self,
+        documents: List[str],
+        **kwargs,
+    ) -> OnnxOutputContext:
+        documents = self._preprocess_queries(documents)
+        self.tokenizer.enable_truncation(max_length=10000)
+        encoded = self.tokenize(documents, **kwargs)
+        input_ids = np.array([[2, 9413] + e.ids[2:] for e in encoded])
```
Comment (joein) [inline fastembed/text/colpali_model.py:75]: magic numbers

File: fastembed/image/colpali_model.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        "model_file": "model.onnx",
+    }
+]
+
+
+class ColpaliImageModel(OnnxImageEmbedding):
+    def _preprocess_onnx_input(
+        self, onnx_input: Dict[str, np.ndarray], **kwargs
+    ) -> Dict[str, np.ndarray]:
+        empty_text_placeholder = np.array([257152] * 1024 + [2, 50721, 573, 2416, 235265, 108])
```
Comment (joein) [inline fastembed/image/colpali_model.py:31]: a lot of magic numbers

File: fastembed/image/colpali_model.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        even_attention_mask = np.array([1] * 1030)
```
Comment (joein) [inline fastembed/image/colpali_model.py:32]: magic numbers

**Review by joein — CHANGES_REQUESTED** [review]

File: fastembed/image/colpali_model.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            image_files = [
+                Image.open(image) if not isinstance(image, Image.Image) else image
+                for image in images
+            ]
+            encoded = self.processor(image_files)
+        onnx_input = self._build_onnx_input(encoded)
+        onnx_input = self._preprocess_onnx_input(onnx_input)
+
+        model_output = self.model.run(None, onnx_input)
+        embeddings = model_output[0].reshape(len(images), *supported_onnx_models[0]["dim"])
```
[the remark this answers is not in the document]
Comment (joein) [reply fastembed/image/colpali_model.py:66]: in general, it's better not to override this function, but hooks it's using

File: fastembed/text/colpali_model.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            texts_query.append(query)
+        return texts_query
+
+    def onnx_embed(
+        self,
+        documents: List[str],
+        **kwargs,
+    ) -> OnnxOutputContext:
+        documents = self._preprocess_queries(documents)
+        self.tokenizer.enable_truncation(max_length=10000)
```
[the remark this answers is not in the document]
Comment (joein) [reply fastembed/text/colpali_model.py:75]: tokenizer settings should better be set in load_onnx_model

File: fastembed/text/colpali_model.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+            texts_query.append(query)
+        return texts_query
+
+    def onnx_embed(
+        self,
+        documents: List[str],
+        **kwargs,
+    ) -> OnnxOutputContext:
+        documents = self._preprocess_queries(documents)
```
Comment (joein) [inline fastembed/text/colpali_model.py:74]: feels like it should be done in `embed` method rather than `onnx_embed`

**Review by joein — CHANGES_REQUESTED** [review]

File: fastembed/image/colpali_model.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        },
+        "additional_files": ["model.onnx_data"],
+        "model_file": "model.onnx",
+    }
+]
+
+
+class ColpaliImageModel(OnnxImageEmbedding):
+    empty_text_placeholder = np.array([257152] * 1024 + [2, 50721, 573, 2416, 235265, 108])
+    even_attention_mask = np.array([1] * 1030)
```
[the remark this answers is not in the document]
Comment (I8dNLo) [reply fastembed/image/colpali_model.py:29]: The first one- are augmentation tokens added to model anyway. So we do not need to process them with tokenizer as far as we already know tokenization for it. 
The second one- is a mask of empty text input for image mask. It does not mean anything, it's just [SEP] tokens or something. So we still do not need to compute it. I could rename variables for example

File: fastembed/image/image_embedding.py [hunk 9]
```diff
@@ -5,10 +5,11 @@
 from fastembed.common import ImageInput, OnnxProvider
 from fastembed.image.image_embedding_base import ImageEmbeddingBase
 from fastembed.image.onnx_embedding import OnnxImageEmbedding
+from fastembed.image.colpali_model import ColpaliImageModel
 
 
 class ImageEmbedding(ImageEmbeddingBase):
-    EMBEDDINGS_REGISTRY: list[Type[ImageEmbeddingBase]] = [OnnxImageEmbedding]
+    EMBEDDINGS_REGISTRY: list[Type[ImageEmbeddingBase]] = [OnnxImageEmbedding, ColpaliImageModel]
```
[the remark this answers is not in the document]
Comment (I8dNLo) [reply fastembed/image/image_embedding.py:13]: I'm slightly getting difference with LateInteraction and just text, that's why it happens.

File: fastembed/text/colpali_model.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    }
+]
+
+
+class ColpaliTextModel(OnnxTextEmbedding):
+    query_prefix = "Query: "
+    bos_token = "<s>"
+    pad_token = "<pad>"
+    query_tokens = [2, 9413]
+    image_placeholder_size = (3, 448, 448)
```
[the remark this answers is not in the document]
Comment (I8dNLo) [reply fastembed/text/colpali_model.py:34]: Just use 'TextEmbedding' and it goes brrrr.

File: fastembed/text/colpali_model.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (I8dNLo) [reply fastembed/text/colpali_model.py:75]: 1. Here I should remove max_length to maintain original behaviour as far as it's ignored in original preprocessor
2. Let's discuss, I'm slightly getting it

File: fastembed/text/colpali_model.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
+from typing import Any, Dict, Iterable, List
+
+import numpy as np
+
+from fastembed.common.onnx_model import OnnxOutputContext
+from fastembed.text.onnx_embedding import OnnxTextEmbedding
+
+supported_onnx_models = [
+    {
+        "model": "akshayballal/colpali-v1.2-merged",
```
[the remark this answers is not in the document]
Comment (I8dNLo) [reply fastembed/text/colpali_model.py:10]: It's not really separated in our quantization. We could re-quantize it and separate, but I can not say it will have original behaviour. It could, but I'm not sure 100%

File: fastembed/image/colpali_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (I8dNLo) [reply fastembed/image/colpali_model.py:66]: Could you show a reference?

File: fastembed/image/colpali_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (joein) [reply fastembed/image/colpali_model.py:29]: What I am leading to is that there is currently no way for anyone to figure out what this is and where to look for information when they are looking at the code

File: fastembed/image/colpali_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joein) [reply fastembed/image/colpali_model.py:66]: I mean in a similar way how we override `_preprocess_onnx_input`, `_post_process_onnx_input`, etc.

File: fastembed/text/colpali_model.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (joein) [reply fastembed/text/colpali_model.py:10]: Which image model is used? 
How much does it weight? 
Does it make sense to split the models or the difference is negligible?

File: fastembed/text/colpali_model.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (joein) [reply fastembed/text/colpali_model.py:34]: 2. Models should be separated by the input and output type they produce
- Regular models which produce one embedding per document - `TextEmbedding`
- Regular models which produce one embedding per image - `ImageEmbedding`
- Models which produce sparse vectors for documents - `SparseTextEmbedding`
- Text models which use late interaction mechanism and thus producing a vector per token - `LateInteractionTextEmbedding`
- Image models which use late interaction mechanism and thus producing > 1 embedding per image - `LateInteractionImageEmbedding`

File: fastembed/text/colpali_model.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (joein) [reply fastembed/text/colpali_model.py:75]: if I get it right, you want to set it once and for all, then it does not seem right to have it in the method which is being called for each inference
Instead, it can be set during initialization, e.g. in load_onnx_model:
https://github.com/qdrant/fastembed/blob/adfc03ed0de5f4029409b9f26d6e6d1f80cd11b4/fastembed/late_interaction/colbert.py#L179

File: fastembed/text/colpali_model.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (I8dNLo) [reply fastembed/text/colpali_model.py:10]: Gemma 2B
